devel/apache-arrow/Portfile

```diff
@@ -0,0 +1,209 @@
+
+# Build instructions: https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#build-and-test
+cmake.source_dir    ${worksrcpath}/cpp
+cmake.build_type    Release
+
+set llvm_version    10
+set llvm_prefix     ${prefix}/libexec/llvm-${llvm_version}
+
+cmake.module_path-append \
+                    ${llvm_prefix}/lib/cmake/llvm
+
+depends_build-append \
+                    port:pkgconfig
+
+depends_lib-append \
+                    port:llvm-${llvm_version}
+
+pre-configure {
+    xinstall -d ${workpath}/bin
+    foreach f [glob -nocomplain ${prefix}/bin/ll*-${llvm_version}] {
+        ln -s ${f} ${workpath}/bin/[strsed [file tail ${f}] "s|-${llvm_version}$||"]
+    }
+}
+
+configure.args-append \
+                    -DARROW_FLIGHT=ON \
+                    -DARROW_JEMALLOC=OFF \
+                    -DARROW_ORC=ON \
+                    -DARROW_PARQUET=ON \
+                    -DARROW_PLASMA=ON \
+                    -DARROW_PROTOBUF_USE_SHARED=ON \
+                    -DARROW_PYTHON=ON \
+                    -DARROW_USE_CCACHE=OFF \
+                    -DARROW_WITH_BZ2=ON \
+                    -DARROW_WITH_ZLIB=ON \
+                    -DARROW_WITH_ZSTD=ON \
+                    -DARROW_WITH_LZ4=ON \
+                    -DARROW_WITH_SNAPPY=ON \
+                    -DARROW_WITH_BROTLI=ON \
+                    -DARROW_INSTALL_NAME_RPATH=OFF \
+                    -DLLVM_ROOT=${llvm_prefix}
+
+# used for both apache-arrow and py*-pyarrow
+destroot.target     install
+
+if {${name} eq ${subport}} {
+    # Apache Arrow requires a python3 version to build; use this
+    set python_branch   3.8
+    set python_version  [join [split ${python_branch} .] ""]
+    set python_prefix   ${frameworks_dir}/Python.framework/Versions/${python_branch}
+    set python_bin      ${python_prefix}/bin/python${python_branch}
+
+    depends_build-append \
+                    port:python${python_version} \
+                    port:py${python_version}-cython \
+                    port:py${python_version}-numpy
+
+    depends_lib-append \
+                    port:boost \
+                    port:brotli \
+                    port:bzip2 \
+                    port:c-ares \
+                    port:flatbuffers \
+                    port:gmake \
+                    port:grpc \
+                    port:google-glog \
+                    port:libunwind \
+                    port:lz4 \
+                    port:ninja \
+                    port:openssl \
+                    port:protobuf3-cpp \
+                    port:rapidjson \
+                    port:re2 \
+                    port:snappy \
+                    port:thrift \
+                    port:wget \
+                    port:zlib \
+                    port:zstd
+
+    # Note: Modifying PATH is required to ensure that cmake finds the correct python version
+    configure.env-append \
+                    "PATH=${python_prefix}/bin:${workpath}/bin:$env(PATH)"
```

Comment:
Would it be better to use `-DPYTHON_EXECUTABLE=...` when invoking CMake?

Reply:
I tried that but it doesn't work. See this issue: https://issues.apache.org/jira/browse/ARROW-9532?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel

Reply:
Homebrew has it working, so I'll try it again.

Reply:
Thanks, got that working now. I must've fumbled the earlier attempt.

Reply:
Nope, `py37-pyarrow` fails to build without modifying path, in spite of specifying `-DPYTHON_EXECUTABLE=` in both the `cmake` arguments and `PYARROW_CMAKE_OPTIONS`.

I believe that this is an issue with the upstream `cmake` configuration.

Here are the error logs for the failed py37-pyarrow build: https://paste.z0k.xyz/b822c7df5064
```
--->  Building py37-pyarrow
DEBUG: Executing proc-pre-org.macports.build-build-0
DEBUG: Executing org.macports.build (py37-pyarrow)
DEBUG: Environment: 
CC='/usr/bin/clang'
CC_PRINT_OPTIONS='YES'
CC_PRINT_OPTIONS_FILE='/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/.CC_PRINT_OPTIONS'
CFLAGS='-arch x86_64 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
CPATH='/opt/local/include'
CXX='/usr/bin/clang++'
CXXFLAGS='-arch x86_64 -stdlib=libc++ -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
DEVELOPER_DIR='/Library/Developer/CommandLineTools'
F90FLAGS='-m64'
FCFLAGS='-m64'
FFLAGS='-m64'
LDFLAGS='-arch x86_64'
LIBRARY_PATH='/opt/local/lib'
MACOSX_DEPLOYMENT_TARGET='10.14'
OBJC='/usr/bin/clang'
OBJCFLAGS='-arch x86_64 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
PATH='/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/bin:/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin'
PYARROW_CMAKE_OPTIONS='-DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX="/opt/local" -DCMAKE_INSTALL_NAME_DIR="/opt/local/lib" -DCMAKE_SYSTEM_PREFIX_PATH="/opt/local;/usr" -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_POLICY_DEFAULT_CMP0060=NEW -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_COLOR_MAKEFILE=ON -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_MAKE_PROGRAM=/usr/bin/make -DCMAKE_MODULE_PATH="/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm" -DCMAKE_PREFIX_PATH="/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm" -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_RPATH="/opt/local/lib" -Wno-dev -DARROW_FLIGHT=ON -DARROW_JEMALLOC=OFF -DARROW_ORC=ON -DARROW_PARQUET=ON -DARROW_PLASMA=ON -DARROW_PROTOBUF_USE_SHARED=ON -DARROW_PYTHON=ON -DARROW_USE_CCACHE=OFF -DARROW_WITH_BZ2=ON -DARROW_WITH_ZLIB=ON -DARROW_WITH_ZSTD=ON -DARROW_WITH_LZ4=ON -DARROW_WITH_SNAPPY=ON -DARROW_WITH_BROTLI=ON -DARROW_INSTALL_NAME_RPATH=OFF -DCARES_PREFIX=/opt/local -DgRPC_INSTALL=OFF -DgRPC_BUILD_TESTS=OFF -DgRPC_CARES_PROVIDER=package -DgRPC_ABSL_PROVIDER=package -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_RE2_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DLLVM_ROOT=/opt/local/libexec/llvm-10 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7'
PYARROW_WITH_PARQUET='1'
SDKROOT='/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk'
Executing:  cd "/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build_ext --inplace 
DEBUG: system:  cd "/opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build_ext --inplace 
running build_ext
creating /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build
creating /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build/temp.macosx-10.14-x86_64-3.7
-- Running cmake for pyarrow
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/local -DCMAKE_INSTALL_NAME_DIR=/opt/local/lib -DCMAKE_SYSTEM_PREFIX_PATH=/opt/local;/usr -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_POLICY_DEFAULT_CMP0060=NEW -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_COLOR_MAKEFILE=ON -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_MAKE_PROGRAM=/usr/bin/make -DCMAKE_MODULE_PATH=/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm -DCMAKE_PREFIX_PATH=/opt/local/share/cmake/Modules;/opt/local/libexec/llvm-10/lib/cmake/llvm -DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=ON -DCMAKE_INSTALL_RPATH=/opt/local/lib -Wno-dev -DARROW_FLIGHT=ON -DARROW_JEMALLOC=OFF -DARROW_ORC=ON -DARROW_PARQUET=ON -DARROW_PLASMA=ON -DARROW_PROTOBUF_USE_SHARED=ON -DARROW_PYTHON=ON -DARROW_USE_CCACHE=OFF -DARROW_WITH_BZ2=ON -DARROW_WITH_ZLIB=ON -DARROW_WITH_ZSTD=ON -DARROW_WITH_LZ4=ON -DARROW_WITH_SNAPPY=ON -DARROW_WITH_BROTLI=ON -DARROW_INSTALL_NAME_RPATH=OFF -DCARES_PREFIX=/opt/local -DgRPC_INSTALL=OFF -DgRPC_BUILD_TESTS=OFF -DgRPC_CARES_PROVIDER=package -DgRPC_ABSL_PROVIDER=package -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_RE2_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DLLVM_ROOT=/opt/local/libexec/llvm-10 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 -DPYTHON_EXECUTABLE=/opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7  -DPYARROW_BUILD_CUDA=off -DPYARROW_BUILD_FLIGHT=off -DPYARROW_BUILD_GANDIVA=off -DPYARROW_BUILD_DATASET=off -DPYARROW_BUILD_ORC=off -DPYARROW_BUILD_PARQUET=on -DPYARROW_BUILD_PLASMA=off -DPYARROW_BUILD_S3=off -DPYARROW_BUILD_HDFS=off -DPYARROW_USE_TENSORFLOW=off -DPYARROW_BUNDLE_ARROW_CPP=off -DPYARROW_BUNDLE_BOOST=off -DPYARROW_GENERATE_COVERAGE=off -DPYARROW_BOOST_USE_SHARED=on -DPYARROW_PARQUET_USE_SHARED=on -DCMAKE_BUILD_TYPE=release /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- System processor: x86_64
-- Performing Test CXX_SUPPORTS_SSE4_2
-- Performing Test CXX_SUPPORTS_SSE4_2 - Success
-- Performing Test CXX_SUPPORTS_AVX2
-- Performing Test CXX_SUPPORTS_AVX2 - Success
-- Performing Test CXX_SUPPORTS_AVX512
-- Performing Test CXX_SUPPORTS_AVX512 - Success
-- Arrow build warning level: PRODUCTION
Configured for RELEASE build (set with cmake -DCMAKE_BUILD_TYPE={release,debug,...})
-- Build Type: RELEASE
-- Build output directory: /opt/local/var/macports/build/_Users_runner_work_1_s_devel_apache-arrow/py37-pyarrow/work/arrow-0.17.1/python/build/temp.macosx-10.14-x86_64-3.7/release
CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  version "3.8.5")
Call Stack (most recent call first):
  /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  /opt/local/share/cmake-3.18/Modules/FindPython/Support.cmake:2971 (find_package_handle_standard_args)
  /opt/local/share/cmake-3.18/Modules/FindPython3.cmake:389 (include)
  cmake_modules/FindPython3Alt.cmake:46 (find_package)
  CMakeLists.txt:196 (find_package)


-- Configuring incomplete, errors occurred!
```

Here's are the logs for a successful build of `py38-pyarrow`: https://paste.z0k.xyz/036eb427e525

Reply:
Finally, now it works with `-DPython3_EXECUTABLE=…` thanks to the folks at the arrow jira.